src/main/scala/mesosphere/util/CapConcurrentExecutions.scala

```diff
@@ -0,0 +1,157 @@
+    queued.setValue(0)
+    processing.setValue(0)
+  }
+}
+
+class CapConcurrentExecutions private (
+    metrics: CapConcurrentExecutionsMetrics,
+    actorRefFactory: ActorRefFactory,
+    actorName: String,
+    maxParallel: Int,
+    maxQueued: Int) {
+  import CapConcurrentExecutions.log
+
+  private[this] val serializeExecutionActorProps =
+    RestrictParallelExecutionsActor.props(metrics, maxParallel = maxParallel, maxQueued = maxQueued)
+  private[this] val serializeExecutionActorRef = actorRefFactory.actorOf(serializeExecutionActorProps, actorName)
+
+  def apply[T](block: => Future[T]): Future[T] = {
+    PromiseActor.askWithoutTimeout(
+      actorRefFactory, serializeExecutionActorRef, RestrictParallelExecutionsActor.Execute(() => block))
+  }
+
+  @PreDestroy
+  def close(): Unit = {
+    log.debug(s"stopping $serializeExecutionActorRef")
+    serializeExecutionActorRef ! PoisonPill
+  }
+}
+
+/**
+  * Accepts execute instructions containing functions returning `scala.concurrent.Future`s.
+  * It only allows `maxParallel` parallel executions and queues the other operations.
+  * It will not queue more than `maxQueued` execute instructions.
+  */
+private[util] class RestrictParallelExecutionsActor(
+    metrics: CapConcurrentExecutionsMetrics, maxParallel: Int, maxQueued: Int) extends Actor {
+
+  import RestrictParallelExecutionsActor.{ log, Execute, Queued }
+
+  private[this] var active: Int = 0
+  private[this] var queue: Queue[Queued] = Queue.empty
+
+  override def preStart(): Unit = {
+    super.preStart()
+
+    metrics.reset()
+
+    for (execute <- queue) {
+      execute.sender ! Status.Failure(new IllegalStateException(s"$self actor stopped"))
+    }
+  }
+
+  override def postStop(): Unit = {
+    metrics.reset()
+
+    super.postStop()
+  }
+
+  override def receive: Receive = {
+    case Execute(func) =>
+
+      if (active >= maxParallel && queue.size >= maxQueued) {
+        sender ! Status.Failure(new IllegalStateException(s"$self queue may not exceed $maxQueued entries"))
+      }
+      else {
+        queue :+= Queued(sender(), func)
+        startNextIfPossible()
+      }
+
+    case Finished =>
+      active -= 1
+      startNextIfPossible()
+  }
+
+  private[this] def startNextIfPossible(): Unit = {
+    if (active < maxParallel) {
+      startNext()
+    }
+
+    metrics.processing.setValue(active)
+    metrics.queued.setValue(queue.size)
+  }
+
+  private[this] def startNext(): Unit = {
+    queue.dequeueOption.foreach {
+      case (next, newQueue) =>
+        queue = newQueue
+        active += 1
+
+        import context.dispatcher
+        import akka.pattern.pipe
+        val future: Future[Any] =
+          try metrics.processingTimer.timeFuture(next.func())
+          catch { case NonFatal(e) => Future.failed(e) }
+        future.pipeTo(next.sender)
+
+        val myself = self
+        future.onComplete(_ => myself ! Finished)(CallerThreadExecutionContext.callerThreadExecutionContext)
```

Comment:
To avoid the overhead of yet another context switch. Sending a message to an actor not be (significantly) more expensive than submitting a task to an execution context.